flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/LocalStateForwardingTest.java

```diff
@@ -0,0 +1,213 @@
+
+		performCheck(osFuture.getKeyedStateManagedFuture(), jmState.getManagedKeyedState(), tmState.getManagedKeyedState());
+		performCheck(osFuture.getKeyedStateRawFuture(), jmState.getRawKeyedState(), tmState.getRawKeyedState());
+		performCheck(osFuture.getOperatorStateManagedFuture(), jmState.getManagedOperatorState(), tmState.getManagedOperatorState());
+		performCheck(osFuture.getOperatorStateRawFuture(), jmState.getRawOperatorState(), tmState.getRawOperatorState());
+	}
+
+	/**
+	 * This tests that state that was reported to the {@link org.apache.flink.runtime.state.TaskStateManager} is also
+	 * reported to {@link org.apache.flink.runtime.taskmanager.CheckpointResponder} and {@link TaskLocalStateStore}.
+	 */
+	@Test
+	public void testForwardingFromTaskStateManagerToResponderAndTaskLocalStateStore() {
+
+		final JobID jobID = new JobID();
+		final ExecutionAttemptID executionAttemptID = new ExecutionAttemptID();
+		final CheckpointMetaData checkpointMetaData = new CheckpointMetaData(42L, 4711L);
+		final CheckpointMetrics checkpointMetrics = new CheckpointMetrics();
+		final int subtaskIdx = 42;
+		JobVertexID jobVertexID = new JobVertexID();
+
+		TaskStateSnapshot jm = new TaskStateSnapshot();
+		TaskStateSnapshot tm = new TaskStateSnapshot();
+
+		final AtomicBoolean jmReported = new AtomicBoolean(false);
+		final AtomicBoolean tmReported = new AtomicBoolean(false);
+
+		TestCheckpointResponder checkpointResponder = new TestCheckpointResponder() {
+
+			@Override
+			public void acknowledgeCheckpoint(
+				JobID lJobID,
+				ExecutionAttemptID lExecutionAttemptID,
+				long lCheckpointId,
+				CheckpointMetrics lCheckpointMetrics,
+				TaskStateSnapshot lSubtaskState) {
+
+				Assert.assertEquals(jobID, lJobID);
+				Assert.assertEquals(executionAttemptID, lExecutionAttemptID);
+				Assert.assertEquals(checkpointMetaData.getCheckpointId(), lCheckpointId);
+				Assert.assertEquals(checkpointMetrics, lCheckpointMetrics);
+				jmReported.set(true);
+			}
+		};
+
+		TaskLocalStateStore taskLocalStateStore = new TaskLocalStateStore(jobID, jobVertexID, subtaskIdx) {
+			@Override
+			public void storeLocalState(
+				@Nonnull CheckpointMetaData checkpointMetaData,
+				@Nullable TaskStateSnapshot localState) {
+
+				Assert.assertEquals(tm, localState);
+				tmReported.set(true);
+			}
+		};
+
+		TaskStateManagerImpl taskStateManager =
+			new TaskStateManagerImpl(
+				jobID,
+				executionAttemptID,
+				taskLocalStateStore,
+				null,
+				checkpointResponder);
+
+		taskStateManager.reportTaskStateSnapshots(
+			checkpointMetaData,
+			checkpointMetrics,
+			jm,
+			tm);
+
+		Assert.assertTrue("Reporting for JM state was not called.", jmReported.get());
+		Assert.assertTrue("Reporting for TM state was not called.", tmReported.get());
+	}
+
+	private static <T extends StateObject> void performCheck(
+		Future<SnapshotResult<T>> cc,
+		StateObjectCollection<T> jm,
+		StateObjectCollection<T> tm) {
```

Comment:
Naming of variables could be a bit more expressive.

Reply:
👍